File: Smallest Subarrays With Maximum Bitwise OR/Smallest Subarrays With Maximum Bitwise OR.java
// Solution for Smallest Subarrays With Maximum Bitwise OR in JAVA

class Solution {
    public int[] smallestSubarrays(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        int[] latestBitPosition = new int[32];
        Arrays.fill(latestBitPosition, -1);

        for (int i = n - 1; i >= 0; i--) {
            int farthestReach = i;
            for (int bit = 0; bit < 32; bit++) {
                if ((nums[i] & (1 << bit)) == 0) {
                    if (latestBitPosition[bit] != -1) {
                        farthestReach = Math.max(farthestReach, latestBitPosition[bit]);
                    }
                } else {
                    latestBitPosition[bit] = i;
                }
            }
            result[i] = farthestReach - i + 1;
        }

        return result;
    }
}

File: Smallest Subarrays With Maximum Bitwise OR/Smallest Subarrays With Maximum Bitwise OR.py
# Solution for Smallest Subarrays With Maximum Bitwise OR in PY

class Solution:
    def smallestSubarrays(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0] * n
        latest_bit_position = [-1] * 32

        for i in range(n - 1, -1, -1):
            farthest_reach = i
            for bit in range(32):
                if not (nums[i] & (1 << bit)):
                    if latest_bit_position[bit] != -1:
                        farthest_reach = max(farthest_reach, latest_bit_position[bit])
                else:
                    latest_bit_position[bit] = i

            result[i] = farthest_reach - i + 1

        return result
